generator/app/VerificationBench.hs

```diff
@@ -0,0 +1,275 @@
+    let cfo = defaultConfigurationOptions
+            { cfoFilePath = baConfigPath args
+            , cfoKey = baConfigKey args
+            , cfoSystemStart = Just (Timestamp startTime)
+            }
+    withCompileInfo $
+        withConfigurationsM (LoggerName "verification-bench") Nothing cfo $ \_ !pm ->
+            let tp = TestParams
+                    { _tpStartTime = Timestamp (convertUnit startTime)
+                    , _tpBlockVersionData = genesisBlockVersionData
+                    , _tpGenesisInitializer = genesisInitializer
+                    }
+            in runBlockTestMode tp $ do
+                -- initialize databasea
+                initNodeDBs pm slotSecurityParam
+                bs <- case baBlockCache args of
+                    Nothing -> do
+                        -- generate blocks and evaluate them to normal form
+                        logInfo "Generating blocks"
+                        generateBlocks pm (baBlockCount args)
+                    Just path -> do
+                        fileExists <- liftIO $ doesFileExist path
+                        mbs <- if fileExists
+                                  then liftIO $ readBlocks path
+                                  else return Nothing
+                        case mbs of
+                            Nothing -> do
+                                -- generate blocks and evaluate them to normal form
+                                logInfo "Generating blocks"
+                                bs <- generateBlocks pm (baBlockCount args)
+                                liftIO $ writeBlocks path bs
+                                return bs
+                            Just bs -> return bs
+
+                satisfySlotCheck bs $ do
+                    logInfo "Verifying blocks"
+                    let bss = force $ zip ([1..] :: [Int]) $ replicate (baRuns args) bs
+                    (times, errs) <- fmap unzip $ forM bss
+                        $ \(idx, blocks) -> do
+                            logInfo $ sformat ("Pass: "%int) idx
+                            (if baApply args
+                                then validateAndApply pm blocks
+                                else validate pm blocks)
+
+                    let -- drop first three results (if there are more than three results)
+                        itimes :: [Float]
+                        itimes = map realToFrac (if length times > 3 then drop 3 times else times)
+                        -- execution mean time
+                        mean :: Float
+                        mean = avarage itimes
+                        -- standard deviation of the execution time distribution
+                        stddev :: Float
+                        stddev = sqrt . (\x -> x / realToFrac (length itimes - 1)) . avarage . map ((**2) . (-) mean) $ itimes
+                    logInfo $ sformat ("verification and application mean time: "%shown%"μs stddev: "%shown) mean stddev
```

Comment:
Can we not use criterion for these?

Reply:
We do, this executable I wrote to just gather GHC profiling data. It has a few other options, like reading blocks from a disc cache, so one does not profile block creation (which takes a lot of time).